simplemma/lemmatizer.py

```diff
@@ -185,31 +198,21 @@ def _return_lemma(
     # filters
     if token.isnumeric():
         return token
-    # dictionary search
-    candidate = _simple_search(token, datadict, initial=initial)
-    # simple rules
-    if candidate is None and lang is not None and lang in APPLY_RULES:
-        candidate = APPLY_RULES[lang](token, greedy)
-    # decomposition
-    if candidate is None:  # and greedy is True
-        candidate = _dehyphen(token, datadict, greedy)
-    else:
-        newcandidate = _dehyphen(candidate, datadict, greedy)
-        if newcandidate is not None:
-            candidate = newcandidate
+
+    candidate = (
+        _dehyphen(token, datadict, greedy)
+        or _simple_search(token, datadict, initial=initial)
+        or apply_rules(token, greedy, lang)
+        or _find_prefixes(token, lang, datadict)
+    )
+
     # stop here in some cases
     # if not greedy:
     #    return candidate
     limit = 6 if lang in SHORTER_GREEDY else 8
     if len(token) <= limit:
         return candidate
-    # subword decomposition: predefined prefixes (absent from vocabulary if they are not words)
-    if candidate is None and lang in FIND_KNOWN_PREFIXES:
-        prefix = FIND_KNOWN_PREFIXES[lang](token)
-        if prefix is not None:
-            subword = _simple_search(token[len(prefix) :], datadict)
-            if subword is not None:
-                candidate = prefix + subword.lower()
+
```

Comment:
This is a breaking change. Since words shorter than 6 or 8 characters will know be also tried by the prefix finder.

So, the new version might produce a different result for some tokens.

As you know, I'm happy introducing breaking changes before v1.
But we should think how are we going to handle this in the future.